### PR TITLE
feat!: move built in templates for Go, Node.js and TypeScript to Paketo builders

### PR DESCRIPTION
**Pull request by lance**
# Changes

- :gift: Move builtin templates for Go, Node.js and TypeScript to Paketo builders

/kind enhancement

This commit removes the Boson builders from the built in templates
for Go, TypeScript and Node.js, replacing them with paketo builders.
For Go, the builder is augmented with a simple buildpack that installs
the Go wrapper code and its dependencies. For TypeScript, the paketo
buildpacks oddly don't support an `npm build` step, so these templates
are also dependent on a small Boson buildpack. These buildpacks are
currently at https://github.com/lance/boson-buildpacks but should find
a home either in the boson-project organization, or the knative-sandbox
organization.

This change also slightly modifies how the Node.js and TypeScript
templates are structured, reducing the coupling between the buildpack
and a function project.

This commit includes the code in https://github.com/knative-sandbox/kn-plugin-func/pull/465
and is dependent on it in the use of manifest.yaml.

### CODE REVIEW

File: templates/go/events/manifest.yaml
Comment (lance) [inline templates/go/events/manifest.yaml:8]: Before this PR lands, we should figure out where these should really live - both git org/repo, and container repo.

/hold

File: templates/go/events/manifest.yaml
Comment (lance) [inline templates/go/events/manifest.yaml:2]: Note how all of the templates use the same builder.

File: templates/node/events/manifest.yaml
Comment (lance) [inline templates/node/events/manifest.yaml:5]: Node functions don’t need a special buildpacks.

File: templates/typescript/events/manifest.yaml
Comment (lance) [inline templates/typescript/events/manifest.yaml:8]: Use the paketo buildpacks to install Node.js, npm, and function dependencies. Then use the boson-typescript buildpack to do the TS>JS transpilation.

File: config.go
Comment (lkingland) [inline config.go:140]: Having both a Builder and Builders is perhaps marginally better than appending "Map" to the plural:  Not only is it a bit closer to natural language, it allows us to match `Config` closely with the `Manifest` (one step closer to being an embedded struct).

While there is of course the possibility of confusion, I consider plurality to be an important in establishing member uniqueness.

Throwing caution and potential confusion to the wind, the subtle plurality differentiator is already being embraced with the members `repository` vs `repositories` on the Client.  Perhaps its worth considering here as well.

File: knative/deployer.go
Comment (lkingland) [inline knative/deployer.go:170]: Perhaps not the time for this, but all this indexing into a slice is nerve-wracking.  It's one of those things begging for a future error.

This should probably be refactored to be a single container, because at no point in this function is it used in any way other than a single container, until line 204, at which point would be more appropriate to set it as the only member of slice defined inline anonymously.  The complexity starting to compound as now the `setHealthEndpoints` function needs to pass a reference, making this indexing now pass through a function boundary.

Perhaps better done during a dedicated refactor pass, but thought I'd mention it now, since the function's being touched, that's usually the best time to spruce things up.

File: knative/deployer.go
Comment (lkingland) [inline knative/deployer.go:232]: Agreed... perhaps these values do not need to be populated and passed through for each Function.  We know that 99.9% of the time they will have the default values.

Perhaps the solution here is to make these health endpoint values default _unless_ they exist in the language pack.  That way we are sticking with the general principle:  most everything is invisible to the Function developer using sane defaults... unless one wants to pop the proverbial hood and override things.

By using defaults here in the Deployer, the Function developer would only see these values populated if the language pack chose to populate them... which would _also_ only need to happen if the language pack developer has a reason to deviate from the default.

File: knative/deployer.go
Comment (lkingland) [inline knative/deployer.go:237]: Per my earlier comment, health endpoints may be better solved by providing sane defaults in `deployer.go` directly.  However, this is a good time to drop 2c on the global config.  In general I think this is a very good idea and a good way forward.

We can take the `git` configuration system as a guide here I think:

There are many settings which we default for the Function developer.   Making guesses on their behalf being preferred over prompting for everything.

There exist certain settings where the default chosen presents a nontrivial chance of the user desiring to modify that value.  This can be accomplished by the user modifying their local Function's `.func.yaml` either manually, via the CLI's `config` subcommand, or via the client library if this is an integration situation.

In the event this change to the default value is something the user would like to happen by default on all Functions created by the current user, they can make the change in their global `$XDG_CONFIG_HOME/func/config.yaml` (or via the CLI by providing the `--global` flag on config changes).

This seems like a sane approach; has worked well for git, and since we expect most users to be familiar with git already, this reuses their knowledge; requiring little cognitive load to assimilate.

File: knative/deployer_test.go
Comment (lkingland) [inline knative/deployer_test.go:11]: Fleshing out the idea of my earlier comment,  should it strike your fancy:
Tests would look something like adding an additional test `Test_setHealthEndpointsDefaults` which ensures that a Function with no values in its `.HealthEndpoints` results in the LivenessProbe receiving the default values `DefaultLiveness` and `DeafaultReadiness`...  which are of course set with the standard Kube values.

File: templates/go/events/manifest.yaml
Comment (lkingland) [reply templates/go/events/manifest.yaml:2]: That's a nice simplification!

File: templates/go/events/manifest.yaml
Comment (lkingland) [reply templates/go/events/manifest.yaml:8]: It seems to me at least that git's boson-project is the right place for the code, and docker.io is still the most common public registry.

File: templates/go/events/manifest.yaml
Comment (lkingland) [inline templates/go/events/manifest.yaml:12]: If these defaults were instead defined as constants `DefaultLiveness` and `DefaultReadiness` in `deployer.go`, they could be omitted from the default embedded templates.. and from any Function project for that matter, unless the developers (of either Function or Language Pack) specifically need to override.

**Review by lkingland — COMMENTED** [review]
This sure is nice to see coming together!

My only substantive comment (detailed below) is that in general all the default values should probably be yanked out of the `.yaml`s and instead be Go constants. 

In addition, a couple tests would need to be added to ensure that the defaults were indeed being applied.

/approve
(Dear robot: I want to APPROVE this PR as being sufficiently excellent to merge, but not actually TRIGGER a merge in case Lance wants to take some of my suggestions....  🤨 )

File: client.go
Comment (lance) [reply client.go:406]: Makes sense to me

File: config.go
Comment (lance) [reply config.go:140]: > While there is of course the possibility of confusion, I consider plurality to be sufficient for establishing uniqueness, despite the diminutive size of the poor little "s".

:rofl: I agree and will change it

File: templates/go/events/manifest.yaml
Comment (lance) [reply templates/go/events/manifest.yaml:12]: The only issue with this is that it does not allow a language pack provider to forego supplying these endpoints. It places a requirement on all language packs to either provide these or simply specify `/` as the health check endpoint.

Comment (lance) [reply templates/go/events/manifest.yaml]: See the commit message where this is addressed. I suggest we document that language packs wishing to avoid implementing these health checks at `/health/readiness` and `/health/liveness` could use the root path with query parameters. E.g. `/?health=readiness`.

**Review by lkingland — APPROVED** [review]
Lookin good

**Comment by lance** [comment]
/hold
Updating the buildpack images and such

**Comment by lance** [comment]
/hold cancel

Looks like this is fine to go ahead. We will need to update the two buildpack URIs when that has resolved in the boson-project/packs repo.

**Pull request merged**
